# K-fold Cross Validation

- 모든 데이터셋을 Train에 활용할 수 있음
- 정확도를 향상시킬 수 있으며 데이터 부족으로 인한 과소적합을 방지할 수 있다.

- 모든 데이터셋을 validation에 활용할 수 있음.
- 평가에 사용되는 데이터 편중을 막을 수 있으며 결과에 따라 더 일반화된 모델을 구축할 수 있다.

Train set / Valid set / Test set -> K-fold cross validation
- valid set은 Train set으로 부터 추출한다
- n개의 폴드 세트로부터 추출된 예측 평가는 평균해서 평가 결과에 반영하면 된다.

#### ------------Train set -------------------------------------------------------------------------------------------------Test set--------------

- split 1    Fold 1   Fold 2   Fold 3   Fold 4   Fold 5  --->  Fold 1(valid) / 나머지 학습(Train) ---> prediction(split 1)
- split 2    Fold 1   Fold 2   Fold 3   Fold 4   Fold 5  --->  Fold 2(valid) / 나머지 학습(Train) ---> prediction(split 2)
- split 3    Fold 1   Fold 2   Fold 3   Fold 4   Fold 5  --->  Fold 3(valid) / 나머지 학습(Train) ---> prediction(split 3)
- split 4    Fold 1   Fold 2   Fold 3   Fold 4   Fold 5  --->  Fold 4(valid) / 나머지 학습(Train) ---> prediction(split 4)
- split 5    Fold 1   Fold 2   Fold 3   Fold 4   Fold 5  --->  Fold 5(valid) / 나머지 학습(Train) ---> prediction(split 5)

# 1. K-Fold

- sklearn.model_selection의 KFold 클래스

- Decisiontreeclassifier를 iris데이터 적용 후 교차검증 실시
- 의사결정나무에 대해서는 추후 자세히 다룸

In [12]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split, KFold
import numpy as np

iris = load_iris()
X = iris.data
y = iris.target

# 의사결정나무 모델 생성
dt_clf = DecisionTreeClassifier(random_state=156)

# 5개의 폴드 세트로 분리하는 KFold 객체와 폴드 세트별 정확도를 담을 리스트 객체 생성
kfold=KFold(n_splits=5)    # n_splits의 default 값은 5
cv_accuracy = []
print('붗꽃 데이터 세트 크기 :', X.shape[0])

붗꽃 데이터 세트 크기 : 150


- 데이터의 크기가 150개이기 때문에 학습용(train)은 120개, 검증용(valid)은 30개로 분할된다.
- kfold의 split() 메서드를 사용해 학습용(train)/검증용(valid) 데이터를 분할한다.
- split() 메서드를 사용한 데이터 추출은 반환된 인덱스 기반으로 수행해야 한다.
      -> 폴드 별 학습용, 검증용 데스트의 로우 인덱스를 array로 반환한다.

In [23]:
n_iter = 0

# kfold 객체의 split()를 호출하면 폴드 별 학습용, 검증용 데스트의 로우 인덱스를 array로 반환한다.
for train_index, test_index in kfold.split(X):
    # kfold.spli()으로 반환된 인덱스를 이용해 학습용, 검증용 테스트 데이터를 추출한다.
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]
    # 학습 및 예측
    dt_clf.fit(X_train,y_train)
    pred=dt_clf.predict(X_test)
    n_iter+=1
    # 반복 시마다 정확도 측정
    accuracy = np.round(accuracy_score(y_test,pred),4)                # accuracy_score(실제값, 모델 출력값)
    train_size = X_train.shape[0]
    test_size = X_test.shape[0]
    print('\n# {0} 교차 검증 정확도:{1}, 학습데이터 크기:{2}, 검증데이터 크기:{3}'.format(n_iter,accuracy,train_size,test_size))
    print('\n# {0} 검증 인덱스:{1}'.format(n_iter,test_index))
    cv_accuracy.append(accuracy)
    
# 개별 iteration별 정확도를 합하여 평균 정확도 계산
print('\n## 평균 검증 정확도:',np.round(np.mean(cv_accuracy),2))


# 1 교차 검증 정확도:1.0, 학습데이터 크기:120, 검증데이터 크기:30

# 1 검증 인덱스:[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29]

# 2 교차 검증 정확도:0.9667, 학습데이터 크기:120, 검증데이터 크기:30

# 2 검증 인덱스:[30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53
 54 55 56 57 58 59]

# 3 교차 검증 정확도:0.8667, 학습데이터 크기:120, 검증데이터 크기:30

# 3 검증 인덱스:[60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83
 84 85 86 87 88 89]

# 4 교차 검증 정확도:0.9333, 학습데이터 크기:120, 검증데이터 크기:30

# 4 검증 인덱스:[ 90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119]

# 5 교차 검증 정확도:0.7333, 학습데이터 크기:120, 검증데이터 크기:30

# 5 검증 인덱스:[120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137
 138 139 140 141 142 143 144 145 146 147 148 149]

## 평균 검증 정확도: 0.9


# 2. Stratified K-Fold
- 분류에 적용한다.
- 불균형한 분포도를 가진 레이블(종속변수) 데이터 집합을 위한 k폴드 방식(이는 곧 한쪽으로 치우친 분포를 의미한다)
- 위와 같이 불균형한 분포는 학습/검증 데이터 집합에 균등하게 배분되지 못하는 문제가 발생하는데 stratified k폴드는 이를 해결해 줌
- sklearn.model_selection의 StratifiedKFold 클래스

### 먼저 k-fold를 사용해 발생 가능한 문제를 확인하자.

In [24]:
import pandas as pd

iris=load_iris()
iris_df = pd.DataFrame(iris, columns=iris.feature_names)
iris_df['label']=iris.target
iris_df['label'].value_counts()

0    50
1    50
2    50
Name: label, dtype: int64

### 아래 결과를 보면 학습/검증용으로 분할된 데이터는 각 다른 속성을 갖기 때문에 예측에 전혀 이용될 수 없다.

In [38]:
# 3개의 폴드를 생성 후 교차 검증시 생성되는 학습/검증 레이블 데이터 값의 분포도 확인

kfold=KFold(n_splits=3)
n_iter=0
for train_index, test_index in kfold.split(iris_df):
    n_iter +=1
    label_train = iris_df.label.iloc[train_index]
    label_test = iris_df.label.iloc[test_index]
    print('\n## 교차검증 : {0}'.format(n_iter))
    print('-'*40)
    print('학습 레이블 데이터 분포:\n',label_train.value_counts())
    print('검증 레이블 데이터 분포:\n',label_test.value_counts())


## 교차검증 : 1
----------------------------------------
학습 레이블 데이터 분포:
 1    50
2    50
Name: label, dtype: int64
검증 레이블 데이터 분포:
 0    50
Name: label, dtype: int64

## 교차검증 : 2
----------------------------------------
학습 레이블 데이터 분포:
 0    50
2    50
Name: label, dtype: int64
검증 레이블 데이터 분포:
 1    50
Name: label, dtype: int64

## 교차검증 : 3
----------------------------------------
학습 레이블 데이터 분포:
 0    50
1    50
Name: label, dtype: int64
검증 레이블 데이터 분포:
 2    50
Name: label, dtype: int64


### 이제 Stratified K-Fold를 적용하여 문제를 해결해 보자

In [42]:
from sklearn.model_selection import StratifiedKFold

skf = StratifiedKFold(n_splits=3)
n_iter=0

for train_index, test_index in skf.split(iris_df, iris_df.label):
    label_train=iris_df.label.iloc[train_index]
    label_test=iris_df.label.iloc[test_index]
    print('## 교차검증 :{0}'.format(n_iter))
    print('학습 레이블 데이터 분포 :\n',label_train.value_counts())
    print('검증 레이블 데이터 분포 :\n',label_test.value_counts())

## 교차검증 :0
학습 레이블 데이터 분포 :
 2    34
0    33
1    33
Name: label, dtype: int64
검증 레이블 데이터 분포 :
 0    17
1    17
2    16
Name: label, dtype: int64
## 교차검증 :0
학습 레이블 데이터 분포 :
 1    34
0    33
2    33
Name: label, dtype: int64
검증 레이블 데이터 분포 :
 0    17
2    17
1    16
Name: label, dtype: int64
## 교차검증 :0
학습 레이블 데이터 분포 :
 0    34
1    33
2    33
Name: label, dtype: int64
검증 레이블 데이터 분포 :
 1    17
2    17
0    16
Name: label, dtype: int64


### Stratified K-Fold를 적용한 데이터 분리와 예측

In [47]:
df_clf = DecisionTreeClassifier(random_state=156)

skfold=StratifiedKFold(n_splits=3)
n_iter=0
cv_accuracy=[]

# StratifiedKFold는 split() 호출 시 반드시 종속변수 데이터 세트도 같이 입력해야 한다.
for train_index, test_index in skfold.split(X, y):
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]
    # 학습 및 예측
    df_clf.fit(X_train, y_train)
    pred = df_clf.predict(X_test)
    
    # 반복 시마다 정확도 측정
    n_iter +=1
    accuracy = np.round(accuracy_score(y_test,pred),4)
    train_size = X_train.shape[0]
    test_size = X_test.shape[0]
    print('#{0} 교차 검증 정확도:{1}, 학습 데이터 크기:{2}, 검증 데이터 크기:{3}'.format(n_iter,accuracy,train_size,test_size))
    print('#{0} 검증 세트 인덱스:{1}'.format(n_iter, test_index))
    cv_accuracy.append(accuracy)
    
# 교차 검증별 정확도 및 평균 정확도 계산
print('\m## 교차 검증별 정확도 :', np.round(cv_accuracy,4))
print('## 평균 검증 정확도 :', np.mean(cv_accuracy))

#1 교차 검증 정확도:0.98, 학습 데이터 크기:100, 검증 데이터 크기:50
#1 검증 세트 인덱스:[  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  50
  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66 100 101
 102 103 104 105 106 107 108 109 110 111 112 113 114 115]
#2 교차 검증 정확도:0.94, 학습 데이터 크기:100, 검증 데이터 크기:50
#2 검증 세트 인덱스:[ 17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  67
  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82 116 117 118
 119 120 121 122 123 124 125 126 127 128 129 130 131 132]
#3 교차 검증 정확도:0.98, 학습 데이터 크기:100, 검증 데이터 크기:50
#3 검증 세트 인덱스:[ 34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  83  84
  85  86  87  88  89  90  91  92  93  94  95  96  97  98  99 133 134 135
 136 137 138 139 140 141 142 143 144 145 146 147 148 149]
\m## 교차 검증별 정확도 : [0.98 0.94 0.98]
## 평균 검증 정확도 : 0.9666666666666667


# 3. cross_val_score()
- KFold에서 fold 세트를 설정하고 데이터를 분핳하여 예측을 통한 성능 반환을 한번에 수행하는 API
- cross_val_score(estimator, X, y, scoring, cv, ...)
- startified k 폴드 방식으로 데이터를 분할한다.

In [53]:
from sklearn.model_selection import cross_val_score, cross_validate

iris=load_iris()
dt_clf=DecisionTreeClassifier(random_state=156)

X=iris.data
y=iris.target

# 성능 지표는 accurancy, 교차 검증 세트는 3개
score=cross_val_score(dt_clf,X,y,scoring='accuracy',cv=3)
print('교차 검증별 정확도 : {0}'.format(score))
print('평균 검증 정확도 :', np.round(np.mean(score),4))

교차 검증별 정확도 : [0.98 0.94 0.98]
평균 검증 정확도 : 0.9667


# 4. cross_validate()
- cross_val_score()가 하나의 평가 지표만 사용 가능하지만, cross_validate()는 여러 개의 평가지표를 사용할 수 있다.
- 또한 성능 평가 지표와 수행 시간도 함께 제공한다.

- cross_validate(estimator, X, y, scoring=[], cv= , ...)
- scoring에 MSE나 MAE는 앞에 neg_를 붙인다.
- 딕셔너리 형태로 출력됨


- 결과값 확인
             
      - 적합 수행 시간 : x['fit_time']
      - 지표 수행 시간 : x['score_time']
      - 각 지표 출력값 : x['test_평가지표']

# 5. GridSearchCV - 교차 검증과 최적 하이퍼 파라미터 튜닝을 한번에
- sklearn.model_selection의 GridSearchCV 클래스 사용
- 회귀나 분류 같은 알고리즘에 사용되는 하이퍼 파라미터를 순차적으로 입력하면서 편리하게 최적의 파라미터 도출
- for 반복문을 좀 더 간편하게 사용하기 위해 제공되는 알고리즘
- 교차 검증을 기반으로 하이퍼 파라미터 튜닝

- GridSearchCV(estimator, param_grid, scoring, cv, refit=True(default), verbose)
      -> estimator : classifier, regressor, pipeline
      -> param_grid : key + 리스트 값을 갖는 딕셔너리. estimator 튜닝을 위해 파라미터명과 사용될 여러 파라미터값 지정
      -> refit : 최적 하이퍼 파라미터를 찾은 후 estimator에 재학습
      -> GridSearchCV의 verbose : iteration시마다 수행 결과 메시지를 출력합니다.
         1) verbose=0(default)면 메시지 출력 안함
         2) verbose=1이면 간단한 메시지 출력
         3) verbose=2이면 하이퍼 파라미터별 메시지 출력

- 의사결정나무 알고리즘을 적용하는 하이퍼 파라미터 튜닝(해당 하이퍼 파라미터 : max_depth, min_samples_split)
- 하이퍼 파라미터 세트 : 딕셔너리 타입
- 하이퍼 파라미터 명칭 : 문자열 key 값
- 하이퍼 파라미터의 값 : 리스트 형식

- best_params_ : GridSearchCV가 튜닝한 최적 파라미터
- best_score_ : GridSearchCV가 튜닝한 최적 파라미터를 사용했을 경우 정확도
- best_estimator_ : GridSearchCV가 튜닝한 최적 파라미터
- cv_results : GridSearchCV가 출력한 값을 반환. 일반적으로 데이터프레임 함수를 적용하여 반환한다

In [54]:
from sklearn.model_selection import GridSearchCV

iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.2, random_state=121)

dtree = DecisionTreeClassifier()

# 파라미터를 딕셔너리 형태로 설정
parameter = {'max_depth':[1,2,3],'min_samples_split':[2,3]}

- gridsearchcv 객체의 fit 메서드를 수행하면 학습 데이터를 cv에 기술된 폴딩 세트로 분할
- 이후 param_grid에 기술된 하이퍼 파라미터를 순차적으로 변경하면서 학습/평가를 수행하고 그 결과를 cv_results_속성에 기록
- cv_results_는 gridsearchcv의 결과 세트로서 딕셔너리 형태로 key 값과 리스트 형태의 value 값을 가짐

In [62]:
# param_grid의 하이퍼 파라미터를 3개의 fold로 나누어 테스트 수행
grid_dtree = GridSearchCV(dtree, param_grid=parameter, cv=3, refit=True)

# iris 학습 데이터로 param_grid의 하이퍼 파라미터를 순차적으로 학습/평가
grid_dtree.fit(X_train, y_train)

# gridsearchcv 결과를 추출해 dataframe으로 변환
score_df = pd.DataFrame(grid_dtree.cv_results_)
score_df[['params','mean_test_score','rank_test_score','split0_test_score','split1_test_score','split2_test_score']]

params  mean_test_score  rank_test_score  \
0  {'max_depth': 1, 'min_samples_split': 2}         0.700000                5   
1  {'max_depth': 1, 'min_samples_split': 3}         0.700000                5   
2  {'max_depth': 2, 'min_samples_split': 2}         0.958333                3   
3  {'max_depth': 2, 'min_samples_split': 3}         0.958333                3   
4  {'max_depth': 3, 'min_samples_split': 2}         0.975000                1   
5  {'max_depth': 3, 'min_samples_split': 3}         0.975000                1   

   split0_test_score  split1_test_score  split2_test_score  
0              0.700                0.7               0.70  
1              0.700                0.7               0.70  
2              0.925                1.0               0.95  
3              0.925                1.0               0.95  
4              0.975                1.0               0.95  
5              0.975                1.0               0.95

- max_depth : 3, min_samples_split : 2  => 하이퍼 파라미터 최적의 값
- params : 수행할 때마다 적용된 개별 하이퍼 파라미터 값
- rank_test_score : 하이퍼 파라미터별로 성능이 좋은 score 순위를 나타냄. 1이 가장 우수한 순위이며 이때의 파라미터가 최적의 파라미터
- mean_test_score : 하이퍼 파라미터별로 CV의 폴딩 테스트 세트에 대해 총 수행한 평가 평균값(높을수록 우수)

In [66]:
# best_params_ / best_score_
print('GridSearchCV 최적 파라미터 :', grid_dtree.best_params_)
print('GridSearchCV 최고 정확도 :', grid_dtree.best_score_)

GridSearchCV 최적 파라미터 : {'max_depth': 3, 'min_samples_split': 2}
GridSearchCV 최고 정확도 : 0.975


In [69]:
# 최적 성능을 나타내고 있는 하이퍼 파라미터를 저장한 후 test_data에 적용한다.

# GridSearchCV의 refit으로 이미 학습된 estimator 반환
estimator = grid_dtree.best_estimator_

# GridSearchCV의 best_estimator_는 이미 최적 학습이 됐으므로 별도 학스비 필요 없음
pred = estimator.predict(X_test)
print('Test set accuracy :{0:.4f}'.format(accuracy_score(y_test,pred)))

Test set accuracy :0.9667
